src/components/loading_indicator.rs

```diff
@@ -0,0 +1,93 @@
+use gtk::glib;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+
+mod imp {
+    use super::*;
+    use gtk::graphene;
+    use std::cell::Cell;
+
+    #[derive(Default)]
+    pub struct LoadingIndicator {
+        pub(super) start_time: Cell<i64>,
+        pub(super) progress: Cell<f64>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for LoadingIndicator {
+        const NAME: &'static str = "ComponentsLoadingIndicator";
+        type Type = super::LoadingIndicator;
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.set_css_name("loadingindicator");
+        }
+    }
+
+    impl ObjectImpl for LoadingIndicator {
+        fn constructed(&self) {
+            self.obj().connect_visible_notify(|widget| {
+                if widget.is_visible() {
+                    widget.imp().start_time.set(widget.time());
+                    widget.add_tick_callback(|widget, _clock| {
+                        widget.queue_draw();
+                        Continue(widget.is_visible())
+                    });
+                }
+            });
+        }
+    }
+
+    impl WidgetImpl for LoadingIndicator {
+        fn measure(&self, _orientation: gtk::Orientation, for_size: i32) -> (i32, i32, i32, i32) {
+            (0, for_size, -1, -1)
+        }
+
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            let widget = self.obj();
+            let size = widget.width() as f32;
+            let bounds = graphene::Rect::new(0.0, 0.0, size, size);
+
+            let context = snapshot.append_cairo(&bounds);
+            let color = widget.color();
+            context.set_source_rgba(
+                color.red() as _,
+                color.green() as _,
+                color.blue() as _,
+                color.alpha() as _,
+            );
+            let half_size = size as f64 / 2.0;
+
+            let pi = std::f64::consts::PI;
+
+            context.set_line_width(2.0);
+
+            let time = widget.time() - self.start_time.get();
+            let shift = (time as f64 / 300000.0) % (2.0 * pi);
```

Comment:
Mh, I think I prefer having the rotation part defined in CSS to use cairo as little as possible. This is what GTK does for their spinner: https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/theme/Default/_common.scss#L237

Reply:
If I rotate texture with css (made both using cairo and conic gradient) it looks bad, I've already discussed that with Alexander Mikhaylenko

I prefer using cairo instead of conic gradient because it allows to get rounded endings with ease